back/db/migrate/20241230165518_remove_initiatives_from_core_views.rb

```diff
@@ -0,0 +1,6 @@
+class RemoveInitiativesFromCoreViews < ActiveRecord::Migration[7.0]
+  def change
+    update_view :idea_trending_infos, version: 3, revert_to_version: 2
+    drop_view :union_posts, revert_to_version: 3
+  end
+end
```

Comment:
Testing locally, this fails to rollback (is not reversible).

I've noticed that we don't always ensure migrations are reversible, so maybe this doesn't matter?

```
PG::UndefinedColumn: ERROR:  column comments.post_id does not exist
LINE 5:          FULL JOIN ( SELECT comments.post_id AS idea_id,
                                    ^
/cl2_back/bin/rails:13:in `<main>'

Caused by:
ActiveRecord::StatementInvalid: PG::UndefinedColumn: ERROR:  column comments.post_id does not exist (ActiveRecord::StatementInvalid)
LINE 5:          FULL JOIN ( SELECT comments.post_id AS idea_id,
                                    ^
/cl2_back/bin/rails:13:in `<main>'

Caused by:
PG::UndefinedColumn: ERROR:  column comments.post_id does not exist (PG::UndefinedColumn)
LINE 5:          FULL JOIN ( SELECT comments.post_id AS idea_id,
                                    ^
/cl2_back/bin/rails:13:in `<main>'
Tasks: TOP => db:rollback
```

Reply:
Damn, well spotted! It's definitely not a good situation, but it seems tricky to fix. This is because the views become invalid when the post relationship is replaced by the idea relationship (but the migration still works), as the relationship is migrated before the views are updated. However, on rollback I guess the order is reversed but the views cannot be reverted because the idea relationship is not yet reverted to the post relationship.

Perhaps we could do this:
1. Migrate the comments by **adding** a new idea relationship, but not removing the post relationship. This could be a slow process, because we want to copy over all idea IDs.
2. Migrate the views. This should work even with rollback, because at this point, we have both the idea relationship and the post relationship.
3. Remove the post relationship.

Although this would probably not result in an error when reverting, I'm not sure if it really counts as "revertible", because the post_id and post_type would be added on revert without any values, and when the idea relationship is reverted all values for this relationship would be lost. This can also be solved, but I feel like it would be a bit too much effort.

What do you think?

Reply:
Not sure. I'll have a think and get back to you on this.